### PR TITLE
Adapt db models to allow quick scanning of breaches for a given email address

### PR DESCRIPTION
**Pull request by nhnt11**
This took forever and ended up being a relatively small PR. I had trouble understanding what Sequelize was doing under the hood with associations and battled several weird exceptions (one of which turned out to be a Promise race that was my fault 😢 ).

I'm going to try and push a few more commits on top of this to use this for the /scan API, and try and fix CI failure.

### CODE REVIEW

File: db/models/subscriber.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  }, {});
+
+  Subscriber.associate = function(models) {
+    Subscriber.hasOne(models.EmailHash);
+  };
+
+  Subscriber.prototype.saveSha1 = async function(models) {
+    const sha1 = crypto.createHash("sha1").update(this.email).digest("hex");
+    const emailHash = await models.EmailHash.findOrCreate( { where: { sha1 }});
+    await this.setEmailHash(emailHash);
```
Comment (groovecoder) [inline db/models/subscriber.js]: is `setEmailHash` a setter method we get for free with the association?

File: db/models/subscriber.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  };
+
+  return Subscriber;
```
Comment (groovecoder) [inline db/models/subscriber.js]: Need to change the `create-user` migration into a `create-subscriber` migration.

File: db/models/emailhash.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+"use strict";
+
+module.exports = (sequelize, DataTypes) => {
+  const EmailHash = sequelize.define("EmailHash", {
+    sha1: DataTypes.STRING,
+  }, {});
+
+  EmailHash.associate = function(models) {
+    EmailHash.belongsToMany(models.Breach, { through: "BreachedHashes" });
```
Comment (groovecoder) [inline db/models/emailhash.js]: The `BreachedHashes` model will need something like the `BreachedUser.notified` flag to keep track of which breached hashes we've notified already.

File: server.js [hunk 4]
```diff
@@ -13,7 +12,6 @@ const OAuthRoutes = require("./routes/oauth");
 const UserRoutes = require("./routes/user");
 
 const app = express();
-app.use(bodyParser.json());
```
Comment (groovecoder) [inline server.js]: I took out the global `json` body parser so that the `scan` route could use the `urlencoded` body parser.

**Comment by groovecoder** [comment]
@nhnt11 I got excited about the new `emailHash.getBreaches` functionality, so I went ahead and fixed the `scan.hbs` to use the real live results ...

image

I also left a bunch of comments and a question about the sequelize models & migrations changes. If you don't have in-depth time this week to tackle those, I can tackle them myself if you like?

**Comment by groovecoder** [comment]
@pdehaan do you know how we can get `htmllint` to ignore the handlebars lines in the `.hbs` views?

**Comment by groovecoder** [comment]
I tried:
```
groovetop:blurts-server lcrouch$ git diff .htmllintrc views/
diff --git a/.htmllintrc b/.htmllintrc
index 4878d10..f3d1b21 100644
--- a/.htmllintrc
+++ b/.htmllintrc
@@ -8,5 +8,6 @@
   "class-style": "dash",
   "id-class-style": "dash",
   "indent-style": "spaces",
-  "indent-width": 2
+  "indent-width": 2,
+  "raw-ignore-regex": "<!-- htmllint ignore -->[^]*?<!-- htmllint unignore -->"
 }
diff --git a/views/home.hbs b/views/home.hbs
index f1f2347..dd6b99d 100644
--- a/views/home.hbs
+++ b/views/home.hbs
@@ -1,4 +1,7 @@
+<!-- htmllint ignore -->
 {{!< default }}
+<!-- htmllint unignore -->
+
     <div class="top-bar">
       <div class="top-bar-left">
         <h1><a href="/">Firefox Breach Alert</a></h1>
diff --git a/views/scan.hbs b/views/scan.hbs
index d960d13..1a26280 100644
--- a/views/scan.hbs
+++ b/views/scan.hbs
@@ -1,4 +1,6 @@
+<!-- htmllint ignore -->
 {{!< default }}
+<!-- htmllint unignore -->

     <div class="top-bar">
       <div class="top-bar-left">
```

But I get:
```
groovetop:blurts-server lcrouch$ htmllint 'views/*.hbs'

[htmllint error in views/home.hbs ] Error: Invalid configuration
[htmllint error in views/scan.hbs ] Error: Invalid configuration
[htmllint] found 0 errors out of 2 files
```

**Comment by pdehaan** [comment]
> do you know how we can get htmllint to ignore the handlebars lines in the .hbs views?

@groovecoder: I don't, but I see a few issues discussing HBS and smarty templates in the htmllint repo, as well as the https://github.com/htmllint/htmllint/wiki/Options#raw-ignore-regex solution that it looks like you've tried. I can try poking at it a bit today when I get a few minutes. If it's blocking for now, I'd say we just disable htmllint until we can find a solution.

File: db/models/subscriber.js (hunk 1 above)
Comment (nhnt11) [reply db/models/subscriber.js]: Yup! [1][2]

[1] http://docs.sequelizejs.com/manual/tutorial/associations.html#hasone
[2] http://docs.sequelizejs.com/manual/tutorial/associations.html#belongs-to-many-associations

File: db/models/emailhash.js (hunk 3 above)
Comment (nhnt11) [reply db/models/emailhash.js]: BreachedHashes isn't really a model, it's just the name of the table that is used to store the association data. I think we should add the notified flag to the Subscriber model.

File: db/models/breacheduser.js [hunk 5]
```diff
@@ -1,13 +0,0 @@
...
-    userId: DataTypes.INTEGER,
-    breachId: DataTypes.INTEGER,
-    notified: DataTypes.DATE,
-  }, {});
-  BreachedUser.associate = function(models) {
-    BreachedUser.belongsTo(models.User);
-    BreachedUser.belongsTo(models.Breach);
-  };
-  return BreachedUser;
-};
```
[the remark this answers is not in the document]
Comment (nhnt11) [reply db/models/breacheduser.js]: Done

File: db/models/emailhash.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+  };
+
+  return EmailHash;
+};
```
[the remark this answers is not in the document]
Comment (nhnt11) [reply db/models/emailhash.js]: Done.

File: db/models/subscriber.js (hunk 2 above)
Comment (nhnt11) [reply db/models/subscriber.js]: Done.

File: db/models/emailhash.js (hunk 3 above)
Comment (nhnt11) [reply db/models/emailhash.js]: I don't understand how to put the notified flag on BreachedHashes. BreachedHashes is simply a table that stores the association between EmailHash and Breach - Sequelize is just letting us name it what we want, and I chose BreachedHashes. And indeed, `console.log(models.BreachedHashes)` outputs `undefined`.

Comment (groovecoder) [reply db/models/emailhash.js]: We would define a model for `BreachedHashes` and include the additional attribute. It's the part of [the belongs-to-many docs]http://docs.sequelizejs.com/manual/tutorial/associations.html#belongs-to-many-associations) that starts with "If you want additional attributes in your join table ..."

Comment (groovecoder) [reply db/models/emailhash.js]: I've also forgotten why we moved the `sha1` field out of `Subscriber` and into its own model? :/ In any case, I'll work on top of all this in my branch to finish the templates. We can always change models anytime before we deploy.

**Pull request merged**